astra/src/main/java/com/slack/astra/bulkIngestApi/BulkIngestS3Producer.java

```diff
@@ -0,0 +1,344 @@
+
+      batchSizeGauge.set(requests.size());
+
+      // Process the batch
+      produceDocuments(requests);
+      lastProcessTime = System.currentTimeMillis();
+    }
+  }
+
+  @Override
+  protected Map<BulkIngestRequest, BulkIngestResponse> produceDocuments(
+      List<BulkIngestRequest> requests) {
+
+    Map<BulkIngestRequest, BulkIngestResponse> responseMap = new HashMap<>();
+    try {
+      // Group all requests by partition
+      Map<Integer, Set<BulkIngestRequest>> requestsByPartition =
+          aggregateRequestsByPartition(requests);
+
+      // Process all the requests for each partition
+      for (Map.Entry<Integer, Set<BulkIngestRequest>> entry : requestsByPartition.entrySet()) {
+
+        int partition = entry.getKey();
+        Set<BulkIngestRequest> partitionRequests = entry.getValue();
+
+        processBatchedRequests(partition, partitionRequests, responseMap);
+      }
+
+      for (Map.Entry<BulkIngestRequest, BulkIngestResponse> entry : responseMap.entrySet()) {
+        BulkIngestRequest key = entry.getKey();
+        BulkIngestResponse value = entry.getValue();
+        if (!key.setResponse(value)) {
+          LOG.warn("Failed to add result to the bulk ingest request, consumer thread went away?");
+          failedSetResponseCounter.increment();
+        }
+      }
+    } catch (Exception e) {
+      LOG.error("Failed to write batch to S3/kafka", e);
+      for (BulkIngestRequest request : requests) {
+        responseMap.put(
+            request,
+            new BulkIngestResponse(
+                0,
+                request.getInputDocs().values().stream().mapToInt(List::size).sum(),
+                e.getMessage()));
+      }
+    }
+    return responseMap;
+  }
+
+  private void processBatchedRequests(
+      int partition,
+      Set<BulkIngestRequest> requests,
+      Map<BulkIngestRequest, BulkIngestResponse> responseMap)
+      throws Exception {
+
+    // Combine all spans from all requests for this partition
+    Map<String, List<Trace.Span>> combinedIndexDocs = new HashMap<>();
+    int totalDocs = 0;
+
+    // Track how many documents each request contributes
+    Map<BulkIngestRequest, Integer> requestDocCounts = new HashMap<>();
+
+    // Combine all requests' spans into combinedIndexDocs
+    for (BulkIngestRequest request : requests) {
+      int requestDocs = 0;
+      for (Map.Entry<String, List<Trace.Span>> indexDoc : request.getInputDocs().entrySet()) {
+        String index = indexDoc.getKey();
+        combinedIndexDocs
+            .computeIfAbsent(index, k -> new ArrayList<>())
+            .addAll(indexDoc.getValue());
+        requestDocs += indexDoc.getValue().size();
+      }
+      totalDocs += requestDocs;
+      requestDocCounts.put(request, requestDocs);
+    }
+    // Serialize and upload combined data for the partition as a single S3 object
+    byte[] compressedData = WALBatchSerializer.serialize(combinedIndexDocs);
```

Comment:
nit: could call this serializedData, since compression is an implementation detail